File: apprentissage.php
<?php
/**
 * Serveur PHP pour l'application d'apprentissage de conjugaison espagnole.
 * Utilise la base commune verbs_data.json et persiste la progression
 * de manière 100% compatible avec la version Python.
 */

// Si le fichier demandé existe sur le disque, on laisse le serveur PHP natif le servir
$uri = $_SERVER['REQUEST_URI'];
$uri = explode('?', $uri)[0]; // Retirer les paramètres GET pour le routage

if (file_exists(__DIR__ . $uri) && !is_dir(__DIR__ . $uri) && $uri !== '/apprentissage.php') {
    return false;
}

// Fichiers de données
$verbs_data_file = __DIR__ . '/verbs_data.json';
$flagged_file = __DIR__ . '/flagged_phrases.json';

// Lire le body JSON une seule fois (php://input ne peut être lu qu'une fois)
$body = null;
if ($_SERVER['REQUEST_METHOD'] === 'POST' || $_SERVER['REQUEST_METHOD'] === 'PUT') {
    $body = json_decode(file_get_contents('php://input'), true);
}

// Résolution dynamique de l'utilisateur actif par paramètre GET/POST ou Cookie (Aline, Pascal, Test)
$active_user = 'Aline';
if (isset($_GET['user'])) {
    $active_user = $_GET['user'];
} elseif (isset($body['user'])) {
    $active_user = $body['user'];
} elseif (isset($_COOKIE['active_user'])) {
    $active_user = $_COOKIE['active_user'];
}
$safe_user = preg_replace('/[^a-zA-Z0-9]/', '', strtolower($active_user));
if (!in_array($safe_user, ['aline', 'pascal', 'camil', 'test'])) {
    $safe_user = 'aline';
}
$stats_file = __DIR__ . '/stats_apprentissage_' . $safe_user . '.json';

// Fonction de nettoyage de clé (accent et espaces), identique à Python
function sanitize_key($raw) {
    if (class_exists('Normalizer')) {
        $normalized = Normalizer::normalize($raw, Normalizer::FORM_D);
    } else {
        // Fallback si l'extension intl n'est pas disponible
        $normalized = strtr(utf8_decode($raw), 
            utf8_decode('àáâãäçèéêëìíîïñòóôõöùúûüýÿÀÁÂÃÄÇÈÉÊËÌÍÎÏÑÒÓÔÕÖÙÚÛÜÝ'), 
            'aaaaaceeeeiiiinooooouuuuyyAAAAACEEEEIIIINOOOOOUUUUY');
        $normalized = utf8_encode($normalized);
    }
    // Retirer les marques de combinaison (Mn)
    $no_accents = preg_replace('/\p{Mn}/u', '', $normalized);
    // Remplacer les espaces par des tirets
    $no_spaces = preg_replace('/\s+/', '-', $no_accents);
    // Retirer tout caractère spécial non conforme
    return preg_replace('/[^a-zA-Z0-9_\-]/', '', $no_spaces);
}

// Fonction de chargement de la progression avec migration automatique
function load_progress($stats_file) {
    if (file_exists($stats_file)) {
        $raw = file_get_contents($stats_file);
        $data = json_decode($raw, true);
        if (!$data) return [];
        
        $pronouns_map = [
            "0" => "yo",
            "1" => "tu",
            "2" => "elella",
            "3" => "nosotros",
            "4" => "vosotros",
            "5" => "ellosellas"
        ];
        
        $migrated = [];
        $dirty = false;
        foreach ($data as $k => $v) {
            $sk = sanitize_key($k);
            
            // Check if key ends with __0, __1, etc.
            $parts = explode("__", $sk);
            if (count($parts) >= 2) {
                $last_part = end($parts);
                if (isset($pronouns_map[$last_part])) {
                    $parts[count($parts) - 1] = $pronouns_map[$last_part];
                    $sk = implode("__", $parts);
                    $dirty = true;
                }
            }
            
            if ($sk !== $k) {
                $dirty = true;
            }
            if (isset($migrated[$sk])) {
                $m_succ = isset($migrated[$sk]["success"]) ? $migrated[$sk]["success"] : 0;
                $v_succ = isset($v["success"]) ? $v["success"] : 0;
                $migrated[$sk]["success"] = $m_succ + $v_succ;

                $m_err = isset($migrated[$sk]["error"]) ? $migrated[$sk]["error"] : 0;
                $v_err = isset($v["error"]) ? $v["error"] : 0;
                $migrated[$sk]["error"] = $m_err + $v_err;

                $m_strk = isset($migrated[$sk]["streak"]) ? $migrated[$sk]["streak"] : 0;
                $v_strk = isset($v["streak"]) ? $v["streak"] : 0;
                $migrated[$sk]["streak"] = max($m_strk, $v_strk);

                if (isset($v["lastDate"])) {
                    if (!isset($migrated[$sk]["lastDate"]) || $v["lastDate"] > $migrated[$sk]["lastDate"]) {
                        $migrated[$sk]["lastDate"] = $v["lastDate"];
                        $migrated[$sk]["lastResult"] = isset($v["lastResult"]) ? $v["lastResult"] : null;
                    }
                }
            } else {
                $migrated[$sk] = $v;
            }
        }
        if ($dirty) {
            file_put_contents($stats_file, json_encode($migrated, JSON_UNESCAPED_UNICODE));
        }
        return $migrated;
    }
    return [];
}

// Helper pour récupérer l'aide de terminaison
function get_hint($hints, $mode, $tense, $pronoun_idx) {
    if (isset($hints[$mode][$tense][$pronoun_idx])) {
        return $hints[$mode][$tense][$pronoun_idx];
    }
    return "";
}

// Enrichir le verbe avec les aides dynamiques
function build_cards($verb_db, $hints, $verb_name) {
    if (!isset($verb_db[$verb_name])) {
        return [];
    }
    $verb = $verb_db[$verb_name];
    $result = [];
    foreach ($verb as $mode => $tenses) {
        $result[$mode] = [];
        foreach ($tenses as $tense => $cards) {
            $enriched = [];
            foreach ($cards as $i => $card) {
                $c = $card;
                if (!isset($c["hint"])) {
                    $c["hint"] = get_hint($hints, $mode, $tense, $i);
                }
                $enriched[] = $c;
            }
            $result[$mode][$tense] = $enriched;
        }
    }
    return $result;
}

// Load list of flagged phrase IDs to ignore
function load_flagged_phrases($flagged_file) {
    if (file_exists($flagged_file)) {
        $raw = file_get_contents($flagged_file);
        $data = json_decode($raw, true);
        if (is_array($data)) {
            return $data;
        }
    }
    return [];
}

// Save list of flagged phrase IDs
function save_flagged_phrases($flagged_file, $flagged_list) {
    file_put_contents($flagged_file, json_encode($flagged_list, JSON_UNESCAPED_UNICODE));
}

// Charger la base de données de verbes commune
if (!file_exists($verbs_data_file)) {
    header('HTTP/1.1 500 Internal Server Error');
    header('Content-Type: text/plain; charset=utf-8');
    echo "Fichier verbs_data.json manquant. Veuillez générer la base de données.";
    exit;
}
$verbs_data_content = file_get_contents($verbs_data_file);
$verbs_data = json_decode($verbs_data_content, true);
if ($verbs_data === null) {
    header('HTTP/1.1 500 Internal Server Error');
    header('Content-Type: text/plain; charset=utf-8');
    echo "Erreur de syntaxe JSON dans verbs_data.json : " . json_last_error_msg();
    exit;
}
$verb_db = isset($verbs_data['verbs_db']) ? $verbs_data['verbs_db'] : [];
$hints = isset($verbs_data['hints']) ? $verbs_data['hints'] : [];

// --- Routage de l'API ---

$action = isset($_GET['action']) ? $_GET['action'] : '';

// 1. API: Liste des verbes et progression globale
if ($action === 'data' || $uri === '/api/data') {
    $payload = [
        "verbs" => array_keys($verb_db),
        "progress" => load_progress($stats_file)
    ];
    header('Content-Type: application/json; charset=utf-8');
    header('Cache-Control: no-store, no-cache, must-revalidate, max-age=0');
    header('Pragma: no-cache');
    echo json_encode($payload, JSON_UNESCAPED_UNICODE);
    exit;
}

// 2. API: Récupérer un verbe enrichi
if ($action === 'verb' || strpos($uri, '/api/verb/') === 0) {
    $verb_name = ($action === 'verb') ? (isset($_GET['name']) ? $_GET['name'] : '') : substr($uri, 10);
    $cards = build_cards($verb_db, $hints, $verb_name);
    header('Content-Type: application/json; charset=utf-8');
    echo json_encode($cards, JSON_UNESCAPED_UNICODE);
    exit;
}

// 3. API: Enregistrer un vote (POST)
if ($action === 'vote' || ($uri === '/api/vote' && $_SERVER['REQUEST_METHOD'] === 'POST')) {
    if (!isset($body['id']) || !isset($body['vote'])) {
        header('HTTP/1.1 400 Bad Request');
        echo json_encode(["status" => "error", "message" => "Champs requis manquants."]);
        exit;
    }
    
    $key = $body['id'];
    $vote = $body['vote']; // 'success', 'error' ou 'reset'
    
    $progress = load_progress($stats_file);
    
    if ($vote === 'reset') {
        $entry = [
            "success" => 0, 
            "error" => 0, 
            "streak" => 0, 
            "lastDate" => null, 
            "lastResult" => null
        ];
    } else {
        $entry = isset($progress[$key]) ? $progress[$key] : [
            "success" => 0, 
            "error" => 0, 
            "streak" => 0, 
            "lastDate" => null, 
            "lastResult" => null
        ];
        $entry[$vote] = (isset($entry[$vote]) ? $entry[$vote] : 0) + 1;
        $entry["lastDate"] = gmdate('Y-m-d\TH:i:s\Z');
        $entry["lastResult"] = $vote;
        
        if ($vote === 'success') {
            $entry["streak"] = (isset($entry["streak"]) ? $entry["streak"] : 0) + 1;
        } else {
            $entry["streak"] = 0;
        }
    }
    
    $progress[$key] = $entry;
    file_put_contents($stats_file, json_encode($progress, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
    
    header('Content-Type: application/json; charset=utf-8');
    echo json_encode($entry, JSON_UNESCAPED_UNICODE);
    exit;
}

// 4. API: Réinitialisation générale (POST)
if ($action === 'reset_all' || ($uri === '/api/reset' && $_SERVER['REQUEST_METHOD'] === 'POST')) {
    file_put_contents($stats_file, json_encode([], JSON_UNESCAPED_UNICODE));
    header('Content-Type: application/json; charset=utf-8');
    echo json_encode(["status" => "ok"], JSON_UNESCAPED_UNICODE);
    exit;
}

// 4b. API: Reset series of day (POST)
if ($action === 'reset_day_series' || ($uri === '/api/reset_day_series' && $_SERVER['REQUEST_METHOD'] === 'POST')) {
    $progress = load_progress($stats_file);
    $today_str = gmdate('Y-m-d');
    
    // Remove all entries from today
    foreach ($progress as $k => $v) {
        if (isset($v["lastDate"]) && substr($v["lastDate"], 0, 10) === $today_str) {
            unset($progress[$k]);
        }
    }
    
    file_put_contents($stats_file, json_encode($progress, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
    header('Content-Type: application/json; charset=utf-8');
    echo json_encode(["status" => "ok"], JSON_UNESCAPED_UNICODE);
    exit;
}

// 5. API: Get flagged phrases list (GET)
if ($uri === '/flagged_phrases.json' || strpos($uri, '/flagged_phrases.json') === 0) {
    $flagged_list = load_flagged_phrases($flagged_file);
    header('Content-Type: application/json; charset=utf-8');
    header('Cache-Control: no-store, no-cache, must-revalidate, max-age=0');
    header('Pragma: no-cache');
    echo json_encode($flagged_list, JSON_UNESCAPED_UNICODE);
    exit;
}

// 6. API: Flag a phrase (POST)
if ($action === 'flag' || ($uri === '/api/flag' && $_SERVER['REQUEST_METHOD'] === 'POST')) {
    $phrase_id = isset($body['id']) ? $body['id'] : null;
    if ($phrase_id) {
        $flagged_list = load_flagged_phrases($flagged_file);
        if (!in_array($phrase_id, $flagged_list)) {
            $flagged_list[] = $phrase_id;
            save_flagged_phrases($flagged_file, $flagged_list);
        }
    }
    header('Content-Type: application/json; charset=utf-8');
    echo json_encode(["status" => "ok"], JSON_UNESCAPED_UNICODE);
    exit;
}

// 7. Accueil (Servir index.html)
if ($uri === '/' || $uri === '/index.php' || $uri === '/index.html' || strpos($uri, '/apprentissage.php') === 0) {
    header('Content-Type: text/html; charset=utf-8');
    readfile(__DIR__ . '/index.html');
    exit;
}

// Si la route n'est pas gérée, on renvoie une 404
header('HTTP/1.1 404 Not Found');
echo "Page non trouvée.";
exit;

File: apprentissage.py
import json, os, http.server, socketserver, sys, re, unicodedata, http.cookies
from datetime import datetime, timezone
sys.path.insert(0, os.path.dirname(__file__))
from verbs_data import VERBS_DB, get_hint

PORT = 8000
HTML_FILE = os.path.join(os.path.dirname(__file__), "index.html")
FLAGGED_FILE = os.path.join(os.path.dirname(__file__), "flagged_phrases.json")

def get_stats_file(headers, path="", body=None):
    cookie_header = headers.get("Cookie", "")
    active_user = "Aline"
    if body and isinstance(body, dict) and "user" in body:
        active_user = body["user"]
    elif "?" in path:
        query_part = path.split("?", 1)[1]
        for pair in query_part.split("&"):
            if "=" in pair:
                k, v = pair.split("=", 1)
                if k == "user":
                    active_user = v
                    break
    elif cookie_header:
        try:
            cookie = http.cookies.SimpleCookie()
            cookie.load(cookie_header)
            if "active_user" in cookie:
                active_user = cookie["active_user"].value
        except Exception:
            pass
    safe_user = "".join([c for c in active_user if c.isalnum()]).lower()
    if safe_user not in ["aline", "pascal", "camil", "test"]:
        safe_user = "aline"
    return os.path.join(os.path.dirname(__file__), f"stats_apprentissage_{safe_user}.json")

def load_flagged_phrases():
    """Load flagged phrases as dict keyed by context (verb__mode__tense__pronoun).
    Auto-migrates old flat list format to new object format."""
    if os.path.exists(FLAGGED_FILE):
        try:
            with open(FLAGGED_FILE, "r", encoding="utf-8") as f:
                data = json.load(f)
            if isinstance(data, list):
                # Migrate old flat array format
                migrated = {"_global": data} if data else {}
                save_flagged_phrases(migrated)
                return migrated
            if isinstance(data, dict):
                return data
        except Exception:
            return {}
    return {}

def save_flagged_phrases(flagged_data):
    """Save flagged phrases object."""
    with open(FLAGGED_FILE, "w", encoding="utf-8") as f:
        json.dump(flagged_data, f, ensure_ascii=False, indent=4)

def get_stats_file(headers, path="", body=None):
    cookie_header = headers.get("Cookie", "")
    active_user = "Aline"
    if body and isinstance(body, dict) and "user" in body:
        active_user = body["user"]
    elif "?" in path:
        query_part = path.split("?", 1)[1]
        for pair in query_part.split("&"):
            if "=" in pair:
                k, v = pair.split("=", 1)
                if k == "user":
                    active_user = v
                    break
    elif cookie_header:
        try:
            cookie = http.cookies.SimpleCookie()
            cookie.load(cookie_header)
            if "active_user" in cookie:
                active_user = cookie["active_user"].value
        except Exception:
            pass
    safe_user = "".join([c for c in active_user if c.isalnum()]).lower()
    if safe_user not in ["aline", "pascal", "camil", "test"]:
        safe_user = "aline"
    return os.path.join(os.path.dirname(__file__), f"stats_apprentissage_{safe_user}.json")

def build_cards(verb_name):
    """Inject hints into verb data before sending to client."""
    verb = VERBS_DB.get(verb_name, {})
    result = {}
    for mode, tenses in verb.items():
        result[mode] = {}
        for tense, cards in tenses.items():
            enriched = []
            for i, card in enumerate(cards):
                c = dict(card)
                if "hint" not in c:
                    c["hint"] = get_hint(mode, tense, i)
                enriched.append(c)
            result[mode][tense] = enriched
    return result

def sanitize_key(raw_key):
    nfd_normalized = unicodedata.normalize('NFD', raw_key)
    no_accents = "".join([c for c in nfd_normalized if unicodedata.category(c) != 'Mn'])
    no_spaces = re.sub(r'\s+', '-', no_accents)
    return re.sub(r'[^a-zA-Z0-9_\-]', '', no_spaces)

def load_progress(file_path):
    if os.path.exists(file_path):
        with open(file_path, "r", encoding="utf-8") as f:
            try:
                data = json.load(f)
            except Exception:
                return {}
        
        pronouns_map = {
            "0": "yo",
            "1": "tu",
            "2": "elella",
            "3": "nosotros",
            "4": "vosotros",
            "5": "ellosellas"
        }
        
        migrated = {}
        dirty = False
        for k, v in data.items():
            sk = sanitize_key(k)
            
            # Check if key ends with __0, __1, etc.
            parts = sk.split("__")
            if len(parts) >= 2 and parts[-1] in pronouns_map:
                parts[-1] = pronouns_map[parts[-1]]
                sk = "__".join(parts)
                dirty = True
                
            if sk != k:
                dirty = True
            if sk in migrated:
                migrated[sk]["success"] = migrated[sk].get("success", 0) + v.get("success", 0)
                migrated[sk]["error"] = migrated[sk].get("error", 0) + v.get("error", 0)
                migrated[sk]["streak"] = max(migrated[sk].get("streak", 0), v.get("streak", 0))
                if v.get("lastDate"):
                    if not migrated[sk].get("lastDate") or v["lastDate"] > migrated[sk]["lastDate"]:
                        migrated[sk]["lastDate"] = v["lastDate"]
                        migrated[sk]["lastResult"] = v.get("lastResult")
            else:
                migrated[sk] = v
        if dirty:
            save_progress(migrated, file_path)
        return migrated
    return {}

def save_progress(progress, file_path):
    with open(file_path, "w", encoding="utf-8") as f:
        json.dump(progress, f, ensure_ascii=False)

class Handler(http.server.SimpleHTTPRequestHandler):
    def log_message(self, *a): pass  # silence logs

    def do_GET(self):
        if self.path == "/":
            with open(HTML_FILE, "rb") as f:
                data = f.read()
            self.send_response(200)
            self.send_header("Content-type", "text/html; charset=utf-8")
            self.end_headers()
            self.wfile.write(data)

        elif self.path.startswith("/api/data"):
            file_path = get_stats_file(self.headers, self.path)
            payload = {
                "verbs": list(VERBS_DB.keys()),
                "progress": load_progress(file_path)
            }
            self._json(payload)

        elif self.path.startswith("/api/verb/"):
            verb = self.path.split("/api/verb/")[1]
            if "?" in verb:
                verb = verb.split("?")[0]
            self._json(build_cards(verb))
        elif self.path == "/conjugaisons.jpg" or self.path.startswith("/conjugaisons.jpg"):
            jpg_file = os.path.join(os.path.dirname(__file__), "conjugaisons.jpg")
            if os.path.exists(jpg_file):
                with open(jpg_file, "rb") as f:
                    data = f.read()
                self.send_response(200)
                self.send_header("Content-type", "image/jpeg")
                self.end_headers()
                self.wfile.write(data)
            else:
                self.send_response(404); self.end_headers()
        elif self.path == "/flag_2.json" or self.path.startswith("/flag_2.json"):
            json_file = os.path.join(os.path.dirname(__file__), "flag_2.json")
            if os.path.exists(json_file):
                with open(json_file, "rb") as f:
                    data = f.read()
                self.send_response(200)
                self.send_header("Content-type", "application/json; charset=utf-8")
                self.end_headers()
                self.wfile.write(data)
            else:
                self.send_response(404); self.end_headers()
        elif self.path == "/flagged_phrases.json" or self.path.startswith("/flagged_phrases.json"):
            flagged_list = load_flagged_phrases()
            self._json(flagged_list)
        else:
            self.send_response(404); self.end_headers()

    def do_POST(self):
        if self.path.startswith("/api/vote"):
            body = json.loads(self.rfile.read(int(self.headers["Content-Length"])))
            file_path = get_stats_file(self.headers, self.path, body)
            progress = load_progress(file_path)
            key = body["id"]
            vote = body["vote"]
            entry = progress.get(key, {"success": 0, "error": 0, "streak": 0, "lastDate": None, "lastResult": None})
            if vote == "reset":
                entry = {"success": 0, "error": 0, "streak": 0, "lastDate": None, "lastResult": None}
            else:
                entry[vote] += 1
                entry["lastDate"] = datetime.now(timezone.utc).isoformat()
                entry["lastResult"] = vote
                if vote == "success":
                    entry["streak"] = entry.get("streak", 0) + 1
                else:
                    entry["streak"] = 0
            progress[key] = entry
            save_progress(progress, file_path)
            self._json(entry)
        elif self.path.startswith("/api/flag"):
            body = json.loads(self.rfile.read(int(self.headers["Content-Length"])))
            phrase_id = body.get("id")
            context = body.get("context", "_global") or "_global"
            if phrase_id:
                flagged_data = load_flagged_phrases()
                if context not in flagged_data:
                    flagged_data[context] = []
                if phrase_id not in flagged_data[context]:
                    flagged_data[context].append(phrase_id)
                    save_flagged_phrases(flagged_data)
            self._json({"status": "ok"})
        elif self.path.startswith("/api/reset"):
            # read user from optional query parameters
            file_path = get_stats_file(self.headers, self.path)
            with open(file_path, "w", encoding="utf-8") as f:
                json.dump({}, f)
            self._json({"status": "ok"})
        elif self.path.startswith("/api/reset_day_series"):
            # Reset progress entries for today only
            file_path = get_stats_file(self.headers, self.path)
            progress = load_progress(file_path)
            today_str = datetime.now(timezone.utc).isoformat()[:10]
            
            # Remove all entries from today
            keys_to_remove = [k for k, v in progress.items() if v.get("lastDate", "")[:10] == today_str]
            for k in keys_to_remove:
                del progress[k]
            
            save_progress(progress, file_path)
            self._json({"status": "ok"})

    def _json(self, obj):
        data = json.dumps(obj, ensure_ascii=False).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-type", "application/json; charset=utf-8")
        self.send_header("Cache-Control", "no-store, no-cache, must-revalidate, max-age=0")
        self.send_header("Pragma", "no-cache")
        self.end_headers()
        self.wfile.write(data)

if __name__ == "__main__":
    socketserver.TCPServer.allow_reuse_address = True
    with socketserver.TCPServer(("", PORT), Handler) as httpd:
        print(f"[OK] http://localhost:{PORT}")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("Arrêt.")
